drivers: rtc: define RTC API driver

nfaurant opened the pull request:
- define RTC api in rtc.h
 - implement rtc driver using internal stm32u5 rtc
 - add rtc sample to test internal stm32u5

erwango reviewed (changes requested):
Thanks for this proposal.

Please the change into several commits:
- API
- Driver implementation
- Sample

Then, I guess we should go into api meeting to see if the proposed api can make it in.
@carlescufi ?

nfaurant committed:
drivers: rtc: Dedicated API proposal

 - Currently all RTC peripherals are implemented through Counter
 - This new API is specific to RTC independently from Counter

Signed-off-by: FAURANT Nicolas <n.faurant@lacroix.group>

nfaurant committed:
drivers: rtc: Driver implementation for stm32u5

 - this illustrates RTC API proposal
 - this driver use stm32 hal API

Signed-off-by: FAURANT Nicolas <n.faurant@lacroix.group>

nfaurant committed:
drivers: rtc: sample to test internal stm32u5 rtc

 - set time
 - get time
 - wake up timer interrupt
 - AlarmA and AlarmB interrupts

Signed-off-by: FAURANT Nicolas <n.faurant@lacroix.group>

erwango commented:
As a reference for further discussion, previous attempt to introduce RTC API: https://github.com/zephyrproject-rtos/zephyr/pull/23526

carlescufi commented:
> Thanks for this proposal.
> 
> Please the change into several commits:
> 
> * API
> * Driver implementation
> * Sample
> 
> Then, I guess we should go into api meeting to see if the proposed api can make it in. @carlescufi ?

Indeed. @nfaurant thank you for raising this subject again. I would recommend that you read through the conclusions that we wrote down the last time we brought this up in the API meeting:
https://github.com/zephyrproject-rtos/zephyr/pull/23526#issuecomment-603363339

And compare the original PR with yours. After that, if you think it's suitable then we can bring it to the meeting again.

nfaurant commented:
> And compare the original PR with yours. After that, if you think it's suitable then we can bring it to the meeting again

The purpose of this new API specific to an RTC driver, is to dissociate the use of an RTC features of the Counter API driver.

Indeed, a use case can require the use of the following 2 functionnalities:
- counting external pulses (via the Counter driver)
- having an RTC time base, and wake up periodically or on programmed alarms (via the RTC driver)

Hence, the interest to separate funcionnalities in 2 drivers

thedjnK commented:
+1 on this driver class, would also add a function for getting number of alarm channels and another to check if the RTC is running or stopped. At present, the 2 RTC drivers using the cou nter driver class have global functions to call to set the time because the counter API doesn't have a way to set it, and it's a very ugly way to use an RTC.

nfaurant commented:
It's easy to add a function for getting number of alarm channel.
Regarding the function to check if the RTC is running or stopped, I think that we can only check if the rtc was initialized and set. 
Maybe you want to check if an alarm or a wakeup timer is starting or stopped ?

thedjnK commented:
Most RTCs allow the clock to be stopped, e.g. rp2040 and mcp7940, so the 'is running' function would be to return a bool if the clock is actively being incremented or not

nfaurant commented:
In this case we could add the following API functions:
rtc_start(const struct device *dev)
rtc_stop(const struct device *dev)
rtc_get_status(const struct device *dev)

lairdjm commented:
@nfaurant are you able to join the API review call in 30 minutes?

nfaurant commented:
Sorry, I was not available. Have you reviewed the RTC API ?

lairdjm commented:
Not in yesterday's meeting, would you be available for the next API review meeting on Tuesday to introduce and give an overview of this RFC/API?

nfaurant commented:
I will be at next API review meeting (2022/03/15)

carlescufi commented:
API meeting:

- The consensus seems to have changed since the original discussion, where the impression was that using `counter.h` combined with conversions to and from civil time would be a good idea
- Agreement is that an RTC API is useful to avoid conversions at the driver or application level, simplifying the overall code complexity, at the cost of some duplication
- Some of the peripherals (like STM32's RTC) might offer both the counter and RTC APIs so that applications can choose what to interface it as
- The new API will be agreed in this PR

de-nordic commented on include/drivers/rtc.h:
```diff
@@ -0,0 +1,262 @@
+__subsystem struct rtc_driver_api {
+	int (*set_time)(const struct device *dev, const struct tm date_time);
+	int (*get_time)(const struct device *dev, struct tm *date_time);
+	int (*set_wakeup_timer)(const struct device *dev, struct rtc_wakeup wut, enum rtc_wakeup_id wut_id);
+	int (*start_wakeup_timer)(const struct device *dev, enum rtc_wakeup_id wut_id);
+	int (*stop_wakeup_timer)(const struct device *dev, enum rtc_wakeup_id wut_id);
```
Shouldn't wake up be separate API? Have external i2c/parallel/spi devices been considered?
You may have devices like https://datasheets.maximintegrated.com/en/ds/DS1371.pdf or https://ww1.microchip.com/downloads/en/DeviceDoc/MCP7940M-Low-Cost%20I2C-RTCC-with-SRAM-20002292C.pdf, connected via i2c, that may use the alarm API but does not have wake up functionality. On the other side it is possible, on some devices, to use watchdog as a wake up device, even though the device itself does not have RTC, only various timers.
Having "wake-up" functionality as mandatory means that a lot of i2c devices will just have -ENOTSUP added just because it is required.

nfaurant replied:
Some external RTC (like NXP PCF85063A) use this feature under the name "Countdown timer" or "programmable timer".
"wakeup timer" is the name used by ST.
I suggest keeping this feature in the RTC API because this timer is specific to the RTC and sometimes it'usefull for ultra low power periodic wake-up. The user is free to use them or not. In this way the API is generic.
Like for the other driver, if an RTC doesn't implement the wakeup timer fonctionnality, the driver can return -ENOTSUP with a log message:
		LOG_ERR("Wake up timer is not supported");
		return -ENOTSUP;

de-nordic replied:
>(...) In this way the API is generic. Like for the other driver, if an RTC doesn't implement the wakeup timer fonctionnality, the driver can return -ENOTSUP with a log message: LOG_ERR("Wake up timer is not supported"); return -ENOTSUP;

Problem with generic APIs that they bring a lot of stuff that you may not be using but will take up your flash anyway.
Separate wake-API would be more generic because it would allow to not only cover this specific RTC tied feature but the mentioned watchdog based or count down based, whatever it is called, when you are not interested in entire RTC thing.

nfaurant replied:
We can add a symbol in the Kconfig of the rtc

config RTC_EN_WUT
	bool "Enable the rtc wake up timer support"
	help
	  This option enables the rtc wake up timer API calls.
	  
And decide to enable or not the wake up timer API
__subsystem struct rtc_driver_api {
	int (*set_time)(const struct device *dev, const struct tm date_time);
	int (*get_time)(const struct device *dev, struct tm *date_time);
#ifdef CONFIG_RTC_EN_WUT
	int (*set_wakeup_timer)(const struct device *dev, struct rtc_wakeup wut, enum rtc_wakeup_id wut_id);
	int (*start_wakeup_timer)(const struct device *dev, enum rtc_wakeup_id wut_id);
	int (*stop_wakeup_timer)(const struct device *dev, enum rtc_wakeup_id wut_id);
#endif
	int (*set_alarm)(const struct device *dev, struct rtc_alarm alarm, enum rtc_alarm_id alarm_id);
	int (*start_alarm)(const struct device *dev, enum rtc_alarm_id alarm_id);
	int (*stop_alarm)(const struct device *dev, enum rtc_alarm_id alarm_id);
};
idem for the wut data ...

In this way, an rtc can implement the wut or not without additional memory constraints ...

de-nordic replied:
I am not sure whether #ifdefing API interface is good idea, I know it is done at least with Flash API, but I remember that there have been some arguments against this approach.

Anyway, your proposal is actually reverse of I was thinking of.
I will describe thing I did some time ago to better let you understand what I have meant:
the device I  have in mind was controlling several other devices and conducting certain tests on them (that could take several minutes) at specified intervals and then would go to sleep; the device had been sleeping most of the time, when not doing its main job, to save power as it could be running on batteries (VRLA, AGM or GEL, on a power failure), and would be woken up by a watchdog at certain intervals (ranging from minutes to hours) depending on previous state. The second wake up channel, external, would be a communication device that would wake up the control device upon receiving request from base.
So both ways to get the device to life would be via wake up channel, one timed and one asynchronous interrupt, in both cases the device would check comm device for any messages to be processed.
As you can see such device would have to get entire API to just use the wake up portion of it.

NickolasLapp replied:
I agree with @de-nordic that this seems better as a separate API to implement in addition to the RTC API, rather than lumped in here with an #ifdef. It feels like it's polluting the core idea of what it means to be an RTC with a (common but specific) implementation type.

nfaurant replied:
Therefore, those 3 functions should be removed form rtc API

thedjnK commented:
@nfaurant Any resolution for the conflict?

NickolasLapp commented on include/drivers/rtc.h:
```diff
@@ -0,0 +1,262 @@
+	RTC_ALARM_WEEKDAY_SEL,  /**< Alarm Weekday is selected, set parameter to a value in the 0-6 range (days since Sunday). */
+};
+
+/** @brief Specifies the RTC Alarm Masks. */
```
A bit more specificity about what this mask means exactly would be useful.
I'm looking at RTC functionality for the [MAX31343](https://datasheets.maximintegrated.com/en/ds/MAX31343.pdf) as a possible use case for this API.

That RTC has the option to mask a variety of time fields (from seconds up to years), but only some combinations are valid. For example, you cannot mask to match on minutes but not on seconds. I _think_ that's implied with this enum but I am not sure without more context.

nfaurant replied:
I test this driver with an internal rtc of a STM32 µC (an example was committed).
For this rtc you can add a mask on
 - secondes => RTC_ALARM_MASK_SEC
 - minutes => RTC_ALARM_MASK_MIN
 - hours => RTC_ALARM_MASK_HOURS
 - date or weekday => RTC_ALARM_MASK_DATE_WEEKDAY
 
 If you select RTC_ALARM_MASK_MIN the alarm will occur every minutes
 if you select RTC_ALARM_MASK_DATE_WEEKDAY the alarm will occur
  - every sunday with alarm_time.tm_wday = 0 and use  alarm_date_weekday_sel=RTC_ALARM_WEEKDAY_SEL
  - every monday with alarm_time.tm_wday = 1 and use  alarm_date_weekday_sel=RTC_ALARM_WEEKDAY_SEL
  ...
  - every 2nd of month with alarm_time.tm_mday = 2 and use  alarm_date_weekday_sel=RTC_ALARM_DATE_SEL

bjarki-trackunit commented on include/drivers/rtc.h:
```diff
@@ -0,0 +1,262 @@
+/** @brief Wake up timer structure. */
+struct rtc_wakeup {
+	uint32_t period;                                /**< Wake up timer period in milliseconds. */
+	void (*callback)(const struct device *dev);     /**< Callback function. Can be NULL. */
```
We should provide a user_data variable with the alarm callback to allow for multiple instances of drivers, or C++ classes, which will share the same static callback for different alarms. They can provide pointers to themselves in the user_data field.

```
void (*callback)(const struct device *dev, void *user_data);   /**< Callback function. Can be NULL. */
```

We should also add a boolean to choose between periodic wakeup timer and singleshot wakeup timer
```
    bool periodic; /**< Set if wakeup timer should be periodic*/
```
Or the inverse if the default is periodic
```
    bool singleshot; /**< Set if wakeup timer should be single shot*/
```

bjarki-trackunit commented on include/drivers/rtc.h:
```diff
@@ -0,0 +1,262 @@
+	 * If WeekDay date is selected, use tm_wday [0 to 6] to set the alarm
+	 */
+	enum rtc_alarm_date_weekday alarm_date_weekday_sel;
+	void (*callback)(const struct device *dev);     /**< Callback function. Can be NULL. */
```
See this comment https://github.com/zephyrproject-rtos/zephyr/pull/41287/files#r902898525

bjarki-trackunit commented:
Using enums to specify wakeup timer id and alarm id is too rigid, as it will require updating the API every time a "better" RTC device (with more alarms or wakeup timers) is supported, and it can easily lead to issues where two parts of an application are setting the same alarm by accident, which is a bug that can be hard to track down.

Instead, an id (or handle), should be returned when using the **alarm_set** or **wakeup_timer_set** functions, this will alert the application if it tries to set more timers/alarms than the RTC supports, and will prevent two+ parts of the application from using the same alarm/timer. 

This will require a new function like "alarm_give" or similar to make it available again, but remove the need for the enums, making it support "infinite" alarms/triggers from the start in a safer manner.

cfriedt commented on include/drivers/rtc.h:
```diff
@@ -0,0 +1,262 @@
+/** @brief Wake up timers enum. */
+enum rtc_wakeup_id {
+	RTC_WUT0,       /**< Wake Up timer 0. */
+	/* Add new item before RTC_WUT_NUM. */
```
This doesn't scale very well - it should not be required to add enumerations to a global header just to index a wakeup id.

bjarki-trackunit replied:
Out of curiosity, have you ever seen a RTC with more than 1 wakeup timer? Seems unnecessary since only the wakeup timer with the shortest period has any influence on wakeup. Since the RTC must be ultra low power, adding another one would just increase power consumption without adding meaningful functionality.

I would remove the ID, unless anyone knows of an RTC with more than one wakeup timer :)

thedjnK replied:
http://ww1.microchip.com/downloads/en/devicedoc/20005010f.pdf

> Dual Programmable Alarms

bjarki-trackunit replied:
The alarms are different from wakeup timers, most RTCs have more than one alarm, but only one wakeup timer if any. The Wakeup timer just counts up to a match value, and then invokes an interrupt, which can be mapped to wake up the device. The point of the timer is to be very low power, and as such, only has this limited capability.
Alarms on the other hand use the date and time to invoke alarms, and require much more power.

The device linked here has to alarms, no wakeup timer http://ww1.microchip.com/downloads/en/devicedoc/20005010f.pdf

cfriedt replied:
Alternately, just use DT chosen, and then the ID is implicit.

cfriedt commented on include/drivers/rtc.h:
```diff
@@ -0,0 +1,262 @@
+enum rtc_alarm_id {
+	RTC_ALARMA,     /**< Alarm A. */
+	RTC_ALARMB,     /**< Alarm B. */
+	/* Add new item before RTC_ALARM_NUM. */
```
Same as above - would we just not use `size_t`?

cfriedt replied:
Better than size_t, we should just use DT chosen and a device handle.

cfriedt commented:
@stephanosio - this is the most recent attempt to define an RTC API, FWIR

cfriedt commented:
@nfaurant - this may become relevant shortly.

yonsch commented:
Any updates? This would be useful for the RP2040

nordicjm commented:
> Any updates? This would be useful for the RP2040

At this point I think someone else needs to take this PR and finish/submit it.

bjarki-trackunit commented:
@yonsch @nordicjm We have a complimentary PR I wrote a while ago https://github.com/zephyrproject-rtos/zephyr/pull/46747, I will adapt it to match the comments and suggestions of this PR and create a new PR for it. I will link it here when ready, we can determine how to move forward from there :) Probably 1-2 days before it is ready

cfriedt commented:
@bjarki-trackunit - I just made some additional comments.

It would really make more sense to pass around a device handle rather than using integer IDs for wakeup or alarm devices. DT chosen can then be used to make selection uniform across systems.

bjarki-trackunit commented:
> @bjarki-trackunit - I just made some additional comments.
> 
> It would really make more sense to pass around a device handle rather than using integer IDs for wakeup or alarm devices. DT chosen can then be used to make selection uniform across systems.

I agree, maybe something like this?
```
chosen {
    zephyr,wakeup-timer = <&alarma>;
};

rtcc: rtcc@4202000 {
    alarma: alarm@4202010 {
    };
    alarmb: alarm@4202020 {
    };
};
```

de-nordic commented:
@bjarki-trackunit @cfriedt I think that we need RFC PR with collected notes and comments from existing approaches to the RTC API.

nordicjm commented:
Note that this specific PR was the one that was discussed and accepted at the API working group a long time ago.

carlescufi commented:
@nfaurant is this something you'd like to continue working on, or would you rather someone else take over?

BenBenSS commented:
Hi, Sir. I am using the HighTEC compiler and using TC275 to control MCP7940M, From the oscilloscope, it can be seen that the slave address was written twice, followed by the data written (0x12, 0x55, 0x08). Why is the slave address written twice? The correct ones should be the slave address, slave register address, and data. Did I configure them incorrectly? Please Help me.

bjarki-trackunit commented:
@BenBenSS Hi, I would recommend you use the [official Discord server](https://discord.com/channels/720317445772017664/720317445772017667) to get help with device driver issues :)
